### ISSUE TITLE
🦠 Model Request: SARS-CoV-2 In Vitro, Touret et al.

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#465]
### Model Title

SARS-CoV-2 In Vitro, Touret et al.

### Publication

Hello @natividadesusana!

As part of your Outreachy contribution, we have assigned you the dataset "SARS-CoV-2 In Vitro, Touret et al." from the Therapeutics Data Commons to try and build a binary classification ML model.
Please copy the provided Google Colab template and use this issue to provide updates on the progress. We'll value not only being able to build the model but also interpreting its results.

**Comment by natividadesusana** [comment]
Hello @GemmaTuron, thank you so much!
I'm getting an error loading my TDC data using the specific assay assigned to me.

Could you please check my progress:
https://colab.research.google.com/drive/1B5_fclOu1ehK8LsEbyGPX3ftIIvGzNhX#scrollTo=1Z52oZ57UH_V

**Comment by EstherIdabor** [comment]
Hi @natividadesusana you have to grant access to your colab, by clicking share

**Comment by natividadesusana** [comment]
Oh, right!
It's done, @EstherIdabor could you try again?
Thanks!

**Comment by Malikbadmus** [comment]
@natividadesusana , the label list you are trying to retrieve from the TDC got the wrong value.

**Comment by EstherIdabor** [comment]
@natividadesusana You have been able to fix that right?,
 I think you should increase your training time rather than just leaving it at the default time which is not usually not enough, and also because the model is already at a disadvantage being exposed to an imbalanced dataset, increasing the time should improve the performance.

**Comment by carcablop** [comment]
Hello @natividadesusana
I see that you have an error when you try to run the confusion matrix. The error is when importing ConfusionMatrizDisplay.
I recommend that you do it like this:
`from sklearn.metrics import ConfusionMatrixDisplay`

And try doing it this way.

```
confusion_matrix_valid= confusion_matrix(Y_test, new_Y_prediction_test)

cfmatrdisp= ConfusionMatrixDisplay(confusion_matrix=confusion_matrix_valid)
```

I saw that you had a small syntax error when you assign a title to a plot. You have it like this: plt.ttle("ConfusionMatrix_validation").
And it is plt.title("ConfusionMatrix_validation")

Since you have a very unbalanced data set, it is very important that you see your confusion matrix, this will help you measure the performance of your model, so you can also make decisions about which metric is more useful for you, and how you could improve the performance.

**Comment by carcablop** [comment]
Another thing I've seen @natividadesusana 
In the roc curve part, I see that you are using the variable "Y_disc_predict_valid", these results of this variable I see that they depend on the threshold. In the case of ROC curves this depends on the probability predictions not on the results of the threshold value.

**Comment by GemmaTuron** [comment]
Hi @natividadesusana ,

Good job, @carcablop, @EstherIdabor  and @Malikbadmus provided some relevant feedback for you to consider. Also look for repeated cells, such as mounting google colab twice.
I'd suggest adding some explanation here of the results you are getting to make it easier to follow your outreachy contribution and closing the thread afterwards.

**Comment by natividadesusana** [comment]
The first step was checking out the [Therapeutics Data Commons](https://tdcommons.ai/) website. They give instructions and guidelines for installing required packages, how to use functions and how to access datasets through the provided APIs.
On google Colab (using the provided `LazyQSAR` notebook template, I installed Therapeutics Data Commons by running `!pip install PyTDC` and then importing tdc using import `tdc`.

<img width="1439" alt="Captura de Tela 2022-10-31 às 19 24 10" src="https://user-images.githubusercontent.com/95102911/199121478-ad82e53e-53c3-4265-882b-d7c823914583.png">

**Comment by natividadesusana** [comment]
This is a preview of the dataset.

<img width="1440" alt="Captura de Tela 2022-10-31 às 19 25 41" src="https://user-images.githubusercontent.com/95102911/199121829-412aae0b-aba9-4967-a12c-bcd2f58dcb1f.png">

**Comment by natividadesusana** [comment]
Then the data has to be split into train, test and validation sets. 

<img width="1440" alt="Captura de Tela 2022-10-31 às 19 32 00" src="https://user-images.githubusercontent.com/95102911/199122434-2baa1ef1-5a33-4bd1-a144-63deba2a7467.png">

**Comment by natividadesusana** [comment]
Number of molecules in dataset for train  is  1039
Number of Molecules in dataset for test is  297
Number of Molecules in dataset for valid is  148
 --------------------------------------------------
The number of inactives in train dataset: 977
The number of actives in train dataset: 62

 Number of inactive in testing dataset: 278
The number of active in testing dataset: 19

 Number of inactive in valid dataset: 141
 Number of active in valid dataset:  7
 --------------------------------------------------

**Comment by natividadesusana** [comment]
Number of molecules in dataset for train  is  1039
Number of Molecules in dataset for test is  297
Number of Molecules in dataset for valid is  148
____________________________________________
The number of inactives in train dataset: 977
The number of actives in train dataset: 62

Number of inactive in testing dataset: 278
The number of active in testing dataset: 19

 Number of inactive in valid dataset: 141
 Number of active in valid dataset:  7
 ___________________________________________

**Comment by natividadesusana** [comment]
____________________________________________
Number of molecules in dataset for train  is  1039
Number of Molecules in dataset for test is  297
Number of Molecules in dataset for valid is  148
____________________________________________
The number of inactives in train dataset: 977
The number of actives in train dataset: 62

Number of inactive in testing dataset: 278
The number of active in testing dataset: 19

 Number of inactive in valid dataset: 141
 Number of active in valid dataset:  7
 ___________________________________________

**Comment by natividadesusana** [comment]
Estas são as visualizações para os conjuntos de dados usando matplotlib.

<img width="728" alt="Captura de Tela 2022-10-31 às 19 42 48" src="https://user-images.githubusercontent.com/95102911/199123697-a9781da5-abf3-494b-b44f-2ed01493e6a0.png">

**Comment by natividadesusana** [comment]
These are the visualizations for the datasets using matplotlib.

<img width="723" alt="Captura de Tela 2022-10-31 às 19 44 41" src="https://user-images.githubusercontent.com/95102911/199123884-3aae9d71-9089-4e6b-bdaf-c35bb3c38169.png">

**Comment by natividadesusana** [comment]
Checking molecules to draw an active molecule and an inactive molecule using the RDKIT package.

<img width="1440" alt="Captura de Tela 2022-10-31 às 19 48 47" src="https://user-images.githubusercontent.com/95102911/199124567-a19f5717-48b6-4208-bbba-e63343e0358b.png">

**Comment by natividadesusana** [comment]
Checking active molecules.

<img width="1437" alt="Captura de Tela 2022-10-31 às 19 53 41" src="https://user-images.githubusercontent.com/95102911/199124925-1d0aaab7-4f7a-427b-8587-2206aea51209.png">

**Comment by natividadesusana** [comment]
Selecting random active and inactive molecule.

<img width="1440" alt="Captura de Tela 2022-10-31 às 19 58 55" src="https://user-images.githubusercontent.com/95102911/199125531-57b445a6-0143-426f-a879-28c1f5b38846.png">

**Comment by natividadesusana** [comment]
Using rdkit to draw active and inactive molecules.

<img width="1440" alt="Captura de Tela 2022-10-31 às 20 05 25" src="https://user-images.githubusercontent.com/95102911/199126171-ca0c04d5-d6ac-4df7-bbde-ee9376af6f64.png">

**Comment by natividadesusana** [comment]
These are the ROC curves obtained.

<img width="395" alt="Captura de Tela 2022-10-31 às 20 07 04" src="https://user-images.githubusercontent.com/95102911/199126363-37e3d9af-6764-414b-89df-2deb238c6f87.png">

**Comment by natividadesusana** [comment]
Obtain the precision and recall  of the Validation set.

<img width="1440" alt="Captura de Tela 2022-10-31 às 20 10 41" src="https://user-images.githubusercontent.com/95102911/199126693-5a06ec8a-410c-434e-aec9-79a79f3f2e8c.png">

**Comment by natividadesusana** [comment]
Getting the same metrics for a test suite

<img width="1440" alt="Captura de Tela 2022-10-31 às 20 12 04" src="https://user-images.githubusercontent.com/95102911/199126824-403fd98a-ba4e-4a9e-9d2f-38252a25cdfa.png">

**Comment by natividadesusana** [comment]
Obtendo a precisão e a recuperação do conjunto de validação.

<img width="1437" alt="Captura de Tela 2022-10-31 às 20 15 06" src="https://user-images.githubusercontent.com/95102911/199127134-804bfa43-f72c-4472-9ef8-0aac536a0091.png">

**Comment by natividadesusana** [comment]
Obtaining the precision and recall of the Validation set.

<img width="1437" alt="Captura de Tela 2022-10-31 às 20 15 06" src="https://user-images.githubusercontent.com/95102911/199127134-804bfa43-f72c-4472-9ef8-0aac536a0091.png">

**Comment by natividadesusana** [comment]
Hi @GemmaTuron ,
Could you please help me, an error keeps occurring when getting a contingency table with the validation set.
Thanks.

<img width="882" alt="Captura de Tela 2022-10-31 às 20 17 02" src="https://user-images.githubusercontent.com/95102911/199127306-7f932205-3e02-4ed7-a41c-52d2ff8604be.png">

**Comment by ZakiaYahya** [comment]
Hi @natividadesusana it seems like you didn't define cm anywhere in your code, that's why it giving you this error. Try to fix this using 
from sklearn import metrics
contigency_table = metrics.confusion_matrix(Y_valid, Y_disc_predict_valid)
cm_display = metrics.ConfusionMatrixDisplay(confusion_matrix = contigency_table, display_labels = ["Not Toxic", "Toxic"])
cm_display.plot()
plt.title("ConfusionMatrix_validation")
plt.savefig("ConfusionMatrix_validation.png")
plt.show()

Hope it will work for you.

**Comment by DhanshreeA** [comment]
Hi @natividadesusana just wanna add a couple of things to @ZakiaYahya 's very useful answer. Perhaps what you wanted to do was something like this:
```python
from sklearn.metrics import ConfusionMatrixDisplay as cm
```
Moreoever, I think the import for `from sklearn.metrics._plot.confusion_matrix` is unnecessary. It looks like an internal routine that you likely do not need. 


> Hi @natividadesusana it seems like you didn't define cm anywhere in your code, that's why it giving you this error. Try to fix this using from sklearn import metrics contigency_table = metrics.confusion_matrix(Y_valid, Y_disc_predict_valid) cm_display = metrics.ConfusionMatrixDisplay(confusion_matrix = contigency_table, display_labels = ["Not Toxic", "Toxic"]) cm_display.plot() plt.title("ConfusionMatrix_validation") plt.savefig("ConfusionMatrix_validation.png") plt.show()
> 
> Hope it will work for you.

**Comment by natividadesusana** [comment]
Hello @ZakiaYahya,
That was great and it worked!
Thank you!

**Comment by natividadesusana** [comment]
Hello @DhanshreeA,
That worked!!
Thank you!!

**Comment by natividadesusana** [comment]
Getting a contingency table with the validation set.

<img width="1439" alt="Captura de Tela 2022-11-01 às 13 33 45" src="https://user-images.githubusercontent.com/95102911/199286616-1fc316a9-c96c-4971-ae30-71bd187e6bf1.png">

**Comment by carcablop** [comment]
Hello @natividadesusana.
I suggest you check your roc curve, I see that it is wrong. The roc curve does not depend on the threshold value, you should not do those calculations for the roc curve or to calculate the AUROC, just pass to that function the results you got from the prediction of the probabilities ("predict_prob). You should do it as follows form:
`fpr,tpr,_=roc_curve(Y_valid,Y_predict_valid).`
As you can see, the variable "Y_predict_valid" is the one you get as a result of making the predictions. Here you do not apply a threshold.
Then just call the function to calculate the auroc, you can do it like this:
`print("AUROC", auc(fpr, tpr))`

Remember to import the necessary packages:
`from sklearn.metrics import roc_curve, AUC`
And finally, you draw the graph.

The values ​​you get when you apply the threshold are what you should use for the confusion matrix. Apply a single threshold value, it can be 0.5, and you can play with those values, until that allows you to make a good analysis of your results.

**Comment by GemmaTuron** [comment]
Hello @natividadesusana !

Thanks for your contributions to Ersilia during the Outreachy Application Period! We hope you have learnt and enjoyed as much as the Ersilia team did!
I will close this issue as this was part of the application period solely.
